File: backend/src/main/java/com/exactaworks/testejava/model/Spent.java
package com.exactaworks.testejava.model;

import java.io.Serializable;
import java.time.OffsetDateTime;
import java.util.HashSet;
import java.util.Set;

import javax.persistence.CollectionTable;
import javax.persistence.Column;
import javax.persistence.ElementCollection;
import javax.persistence.Entity;
import javax.persistence.FetchType;
import javax.persistence.GeneratedValue;
import javax.persistence.GenerationType;
import javax.persistence.Id;

@Entity
public class Spent implements Serializable{
	private static final long serialVersionUID = 1L;
	
	@Id
	@GeneratedValue(strategy = GenerationType.AUTO)
	@Column(name = "id", unique = true, updatable = false)
	private Long id;	
	
	@Column(name = "parson_name", nullable = false)
	private String personName;
	
	@Column(name = "description", nullable = false)
	private String description;

	@Column(name = "datetime", nullable = false)
	private OffsetDateTime datetime;
	
	@Column(name = "value", nullable = false)
	private Double value;
	
	@ElementCollection(fetch = FetchType.LAZY)
	@CollectionTable(name = "tags")
	@Column(name="tag")
	private Set<String> tags = new HashSet<>();
	
	public Spent() {}

	public Long getId() {
		return id;
	}

	public void setId(Long id) {
		this.id = id;
	}

	public String getPersonName() {
		return personName;
	}

	public void setPersonName(String personName) {
		this.personName = personName;
	}

	public String getDescription() {
		return description;
	}

	public void setDescription(String description) {
		this.description = description;
	}

	public OffsetDateTime getDatetime() {
		return datetime;
	}

	public void setDatetime(OffsetDateTime datetime) {
		this.datetime = datetime;
	}

	public Double getValue() {
		return value;
	}

	public void setValue(Double value) {
		this.value = value;
	}

	public Set<String> getTags() {
		return tags;
	}

	public void setTags(Set<String> tags) {
		this.tags = tags;
	}
}

File: web/scripts/model/Spent.js
class Spent{
    constructor(_id, _personName, _description, _value, _date){
        Object.assign(this, { _id, _personName, _description, _value })
        this._date = new Date(_date);
        this._tags = [];

        Object.freeze(this);
    }

    getId(){
        return this._id;
    }

    getPersonName(){
        return this._personName;
    }

    getDescription(){
        return this._description;
    }

    getDate(){
        return new Date(this._date.getTime());
    }

    getValue(){
        return this._value;
    }
    
    getTags(){
        return [].concat(this._tags);
    }

    addTag(tag){
        this._tags.push(tag);
    }
    
    addAllTags(tags){
        tags.forEach(tag => {
            this._tags.push(tag);
        })
    }
}